File: Phase1_Modulations/run_modulation_case.py
import math
import numpy as np
from modulation.modulate_signal import modulate_signal
from modulation.demodulate_signal import demodulate_signal
from channel.add_awgn_noise import add_awgn_noise
from metrics.compute_ber import compute_ber


def erfc_array(x):
    x = np.asarray(x, dtype=float)
    return np.vectorize(math.erfc)(x)


def compute_theoretical_ber(ebno_db, spec_type, M):
    ebno = 10 ** (ebno_db / 10)
    if spec_type == "psk":
        # Approximate BER for M-ary PSK in AWGN using exact formula for coherent BPSK/QPSK and an approximation for M>4.
        if M == 2:
            return 0.5 * np.exp(-ebno)
        elif M == 4:
            return 0.5 * np.exp(-ebno)
        else:
            k = np.log2(M)
            return 2 / k * erfc_array(np.sqrt(ebno) * np.sin(np.pi / M))
    elif spec_type == "qam":
        k = np.log2(M)
        q = np.sqrt(M)
        return 4 * (1 - 1 / q) / k * 0.5 * erfc_array(np.sqrt(3 * k * ebno / (2 * (M - 1))))
    return np.full(np.shape(ebno_db), np.nan, dtype=float)


def run_modulation_case(spec, cfg):
    bits_in = np.random.randint(0, 2, size=(cfg["num_bits"],), dtype=np.uint8)
    k = int(np.log2(spec["M"]))

    tx_sig, tx_bits_padded, pad_bits, tx_sym, ref_const = modulate_signal(bits_in, spec)

    rx_bits_clean, rx_sym_clean = demodulate_signal(tx_sig, spec)
    num_orig_bits = bits_in.size
    ber_clean, err_clean = compute_ber(bits_in, rx_bits_clean[:num_orig_bits])

    ber_sim = np.zeros_like(cfg["EbNoVec"], dtype=float)
    err_count = np.zeros_like(cfg["EbNoVec"], dtype=int)
    rx_sig_plot = None
    rx_bits_plot = None
    rx_sym_plot = None

    for ii, EbNo in enumerate(cfg["EbNoVec"]):
        rx_sig, noise_var, snr_db = add_awgn_noise(tx_sig, EbNo, k)
        rx_bits, rx_sym = demodulate_signal(rx_sig, spec)
        ber_sim[ii], err_count[ii] = compute_ber(bits_in, rx_bits[:num_orig_bits])

        if np.isclose(EbNo, cfg["plotEbNo"]):
            rx_sig_plot = rx_sig
            rx_bits_plot = rx_bits
            rx_sym_plot = rx_sym

    ber_theory = compute_theoretical_ber(cfg["EbNoVec"], spec["type"], spec["M"]) if spec["type"] in {"psk", "qam"} else np.full(cfg["EbNoVec"].shape, np.nan)

    return {
        "name": spec["name"],
        "type": spec["type"],
        "M": spec["M"],
        "bitsPerSymbol": k,
        "bitsIn": bits_in,
        "txSig": tx_sig,
        "txSym": tx_sym,
        "txBitsPadded": tx_bits_padded,
        "padBits": pad_bits,
        "refConstellation": ref_const,
        "rxSigPlot": rx_sig_plot,
        "rxBitsPlot": rx_bits_plot,
        "rxSymPlot": rx_sym_plot,
        "berClean": ber_clean,
        "errClean": err_clean,
        "berSim": ber_sim,
        "errCount": err_count,
        "berTheory": ber_theory,
    }

File: Basic_Modulations_MATLAB/run_modulation_case.m
function result = run_modulation_case(spec, cfg)
%RUN_MODULATION_CASE  Simulate one modulation scheme across Eb/N0 values.

bitsIn = randi([0 1], cfg.numBits, 1); % a column vector of random 0s and 1s
k = log2(spec.M);   % Calculates the number of bits per symbol

% Call the modulation function
[txSig, txBitsPadded, padBits, txSym, refConst] = modulate_signal(bitsIn, spec);

% Clean channel
[rxBitsClean, rxSymClean] = demodulate_signal(txSig, spec);
numOrigBits = numel(bitsIn);

% Calculates BER on the clean signal (should be 0)
[berClean, errClean] = compute_ber(bitsIn, rxBitsClean(1:numOrigBits));

% Preallocates arrays to store BER results for each Eb/N0 step
berSim   = zeros(numel(cfg.EbNoVec), 1);
errCount  = zeros(numel(cfg.EbNoVec), 1);
rxSigPlot = [];
rxBitsPlot = [];
rxSymPlot  = [];

% Sweeps through the defined signal-to-noise ratios
for ii = 1:numel(cfg.EbNoVec)
    EbNo = cfg.EbNoVec(ii); % Extract the current Eb/N0 value

    % Add the calculated noise to the transmitted signal
    [rxSig, noiseVar, snrDb] = add_awgn_noise(txSig, EbNo, k);

    % Demodulate the noisy signal
    [rxBits, rxSym] = demodulate_signal(rxSig, spec);

    % Calculates and stores the BER
    [berSim(ii), errCount(ii)] = compute_ber(bitsIn, rxBits(1:numOrigBits));

    if abs(EbNo - cfg.plotEbNo) < 1e-12
        rxSigPlot = rxSig;
        rxBitsPlot = rxBits;
        rxSymPlot  = rxSym;
    end
end

% Theoretical BER (only for standard PSK/QAM cases) using built-in func
berTheory = nan(numel(cfg.EbNoVec), 1);
if spec.type == "psk"
    berTheory = berawgn(cfg.EbNoVec(:), 'psk', spec.M, 'nondiff');
elseif spec.type == "qam"
    berTheory = berawgn(cfg.EbNoVec(:), 'qam', spec.M);
end

result = struct();
result.name           = spec.name;
result.type           = spec.type;
result.M              = spec.M;
result.bitsPerSymbol   = k;
result.bitsIn         = bitsIn;
result.txSig          = txSig;
result.txSym          = txSym;
result.txBitsPadded   = txBitsPadded;
result.padBits        = padBits;
result.refConstellation = refConst;
result.rxSigPlot      = rxSigPlot;
result.rxBitsPlot     = rxBitsPlot;
result.rxSymPlot      = rxSymPlot;
result.berClean       = berClean;
result.errClean       = errClean;
result.berSim         = berSim;
result.errCount       = errCount;
result.berTheory      = berTheory;
end